okhttp-clients/src/main/java/com/palantir/remoting3/okhttp/ConcurrencyLimiters.java

```diff
@@ -0,0 +1,147 @@
+ */
+
+package com.palantir.remoting3.okhttp;
+
+import com.google.common.annotations.VisibleForTesting;
+import com.google.common.cache.CacheBuilder;
+import com.google.common.cache.RemovalCause;
+import com.google.common.util.concurrent.Futures;
+import com.google.common.util.concurrent.ListenableFuture;
+import com.google.common.util.concurrent.SettableFuture;
+import com.netflix.concurrency.limits.Limiter;
+import com.netflix.concurrency.limits.limit.TracingLimitDecorator;
+import com.netflix.concurrency.limits.limit.VegasLimit;
+import com.netflix.concurrency.limits.limiter.DefaultLimiter;
+import com.netflix.concurrency.limits.strategy.SimpleStrategy;
+import com.palantir.remoting3.tracing.okhttp3.OkhttpTraceInterceptor;
+import java.util.Map;
+import java.util.Optional;
+import java.util.Queue;
+import java.util.concurrent.ConcurrentHashMap;
+import java.util.concurrent.ConcurrentMap;
+import java.util.concurrent.LinkedBlockingQueue;
+import okhttp3.Request;
+import org.slf4j.Logger;
+import org.slf4j.LoggerFactory;
+
+final class ConcurrencyLimiters {
+    private static final Logger log = LoggerFactory.getLogger(ConcurrencyLimiters.class);
+    private static final String FALLBACK = "";
+    private final ConcurrentMap<String, AsyncLimiter> limiters = new ConcurrentHashMap<>();
+
+    private final Map<Limiter.Listener, Runnable> activeListeners = CacheBuilder.newBuilder()
+            .weakKeys()
+            .<Limiter.Listener, Runnable>removalListener(notification -> {
+                if (notification.getCause().equals(RemovalCause.COLLECTED)) {
+                    log.warn("Concurrency limiter was leaked."
+                            + " This implies a remoting bug or classpath issue, and may cause degraded performance");
+                    notification.getValue().run();
+                }
+            })
+            .build()
+            .asMap();
+
+    @VisibleForTesting
+    AsyncLimiter limiter(String name) {
+        return limiters.computeIfAbsent(name, key ->
+                new AsyncLimiter(activeListeners, DefaultLimiter.newBuilder()
+                        .limit(TracingLimitDecorator.wrap(VegasLimit.newBuilder().initialLimit(1).build()))
+                        .build(new SimpleStrategy<>())));
+    }
+
+    public AsyncLimiter limiter(Request request) {
+        final String limiterKey;
+        String pathTemplate = request.header(OkhttpTraceInterceptor.PATH_TEMPLATE_HEADER);
+        if (pathTemplate == null) {
+            limiterKey = FALLBACK;
+        } else {
+            limiterKey = request.method() + " " + pathTemplate;
+        }
+        return limiter(limiterKey);
+    }
+
+    static final class AsyncLimiter {
+        private final Map<Limiter.Listener, Runnable> activeListeners;
+        private final Queue<SettableFuture<Limiter.Listener>> waitingRequests = new LinkedBlockingQueue<>();
+        private final Limiter<Void> limiter;
+
+        public AsyncLimiter(
```

Comment:
why do we need to make the limiter asynchronous? from https://github.com/Netflix/concurrency-limits/blob/master/concurrency-limits-core/src/main/java/com/netflix/concurrency/limits/Limiter.java it seems like Limiter#acquire returns immediately?

Reply:
oh i see, nvm.